ostd/src/arch/loongarch/serial.rs

```diff
@@ -0,0 +1,115 @@
+struct Serial {
+    base_address: usize,
+}
+
+impl Serial {
+    const DATA_TRANSPORT_REGISTER_OFFSET: usize = 0;
+    const LINE_STATUS_REGISTER_OFFSET: usize = 5;
+
+    /// Creates a serial driver.
+    ///
+    /// # Safety
+    ///
+    /// The base address must be a valid UART base address.
+    const unsafe fn new(base_address: usize) -> Self {
+        Self { base_address }
+    }
+
+    /// Sends data to the UART.
+    fn send(&self, c: u8) {
+        while !self
+            .line_status_register_flags()
+            .contains(LineStatusRegisterFlags::TFE)
+        {
+            core::hint::spin_loop();
+        }
+        let ptr = self.base_address as *mut u8;
+        // SAFETY: The safety requirements are upheld by the constructor.
+        unsafe {
+            ptr.add(Self::DATA_TRANSPORT_REGISTER_OFFSET)
+                .write_volatile(c);
+        }
+    }
+
+    /// Receives data from the UART.
+    fn recv(&self) -> Option<u8> {
+        if self
+            .line_status_register_flags()
+            .contains(LineStatusRegisterFlags::DR)
+        {
+            let ptr = self.base_address as *mut u8;
+            // SAFETY: The safety requirements are upheld by the constructor.
+            Some(unsafe {
+                ptr.add(Self::DATA_TRANSPORT_REGISTER_OFFSET)
+                    .read_volatile()
+            })
+        } else {
+            None
+        }
+    }
+
+    fn line_status_register_flags(&self) -> LineStatusRegisterFlags {
+        let ptr = self.base_address as *mut u8;
+        // SAFETY: The safety requirements are upheld by the constructor.
+        let c = unsafe { ptr.add(Self::LINE_STATUS_REGISTER_OFFSET).read_volatile() };
+        LineStatusRegisterFlags::from_bits_truncate(c)
+    }
+}
+
+// SAFETY: For correctness purposes, the UART registers should not be accessed concurrently.
+// However, doing so will not cause memory safety violations.
+unsafe impl Send for Serial {}
+unsafe impl Sync for Serial {}
+
+/// The console UART.
+static CONSOLE_COM1: Once<Serial> = Once::new();
+
+/// Initializes the serial port.
+pub(crate) fn init() {
+    let chosen = DEVICE_TREE.get().unwrap().find_node("/chosen").unwrap();
+    let stdout_path = chosen.property("stdout-path").unwrap().as_str().unwrap();
+    let uart = DEVICE_TREE.get().unwrap().find_node(stdout_path).unwrap();
+    let base_address = uart.reg().unwrap().next().unwrap().starting_address as usize;
+    // SAFETY: It is safe because the UART address is acquired from the device tree,
+    // and be mapped in DMW2.
+    CONSOLE_COM1.call_once(|| unsafe { Serial::new(paddr_to_daddr(base_address)) });
+}
+
+/// Sends a byte on the serial port.
+pub(crate) fn send(data: u8) {
+    // Note: It is the caller's responsibility to acquire the correct lock and ensure sequential
+    // access to the UART registers.
+    if let Some(uart) = CONSOLE_COM1.get() {
```

Comment:
Prefer to use early return to avoid unnecessary indentations.
```rust
    let Some(uart) = CONSOLE_COM1.get() else {
        return;
    };
    match data {
        b'\n' => {
            uart.send(b'\r');
            uart.send(b'\n');
        }
        c => uart.send(c),
    }
```